tests/functional/repl/doc-functor.nix

```diff
@@ -0,0 +1,101 @@
+rec {
+  /**
+    Look, it's just like a function!
+  */
+  multiply = p: q: p * q;
+
+  multiplier = {
+    factor = 2;
+    /**
+      Multiply the argument by the factor stored in the factor attribute.
+    */
+    __functor = self: x: x * self.factor;
+  };
+
+  doubler = {
+    description = "bla";
+    /**
+      Multiply by two. This doc probably won't be rendered because the
+      returned partial application won't have any reference to this location;
+      only pointing to the second lambda in the multiply function.
+    */
+    __functor = self: multiply 2;
+  };
+
+  makeOverridable = f: {
+    /**
+      This is a function that can be overridden.
+     */
+    __functor = self: f;
+    override = throw "not implemented";
+  };
+
+  /**
+    Compute x^2
+   */
+  square = x: x * x;
+
+  helper = makeOverridable square;
+
+  # Somewhat analogous to the Nixpkgs makeOverridable function.
+  makeVeryOverridable = f: {
+    /**
+      This is a function that can be overridden.
+     */
+    __functor = self: arg: f arg // { override = throw "not implemented"; overrideAttrs = throw "not implemented"; };
+    override = throw "not implemented";
+  };
+
+  helper2 = makeVeryOverridable square;
+
+  # The RFC might be ambiguous here. The doc comment from makeVeryOverridable
+  # is "inner" in terms of values, but not inner in terms of expressions.
+  # Returning the following attribute comment might be allowed.
+  # TODO: I suppose we could look whether the attribute value expression
+  #       contains a doc, and if not, return the attribute comment anyway?
```

Comment:
@hsjobeki What do you think?
Currently we follow the [disambiguation rule](https://github.com/NixOS/rfcs/pull/145/files#diff-5050f8573da95e8531bf14e0b0de89d58a680478678ed4f2ed20f9e53def4dd9R200) at the value level.
Was that the intent, or should we only apply it at the expression level, so that in cases like this we can resolve to a more specific comment.
Or perhaps more simply, when the `:doc` query is specifically for an attribute, perhaps we should just prefer the attribute-based lookup?
I'm also considering the option that we should just show everything and treat the rule as a preference or priority.

> ```
> nix-repl> :doc lib.helper3
> Function `__functor`\
>   … defined at /path/to/tests/functional/repl/doc-functor.nix:45:23
> 
> 
> This is a function that can be overridden.
> ```

Reply:
> Or perhaps more simply, when the :doc query is specifically for an attribute, perhaps we should just prefer the attribute-based lookup?

Yes. But functors are polymorph. I would show (maybe concatenate) both, or add an option for the user to decide what he is interested in.

As a user i would expect `:doc lib.helper3` to yield.

```
:doc lib.helper3
->  Compute x^3
```

The property that the function is wrapped with `makeVeryOverridable` is not very interesting.
But i dont know how to solve this issue yet.

Reply:
For example this one edge case that i couldn't solve on a value level, because lamdbas are opaque.

`pkgs.fetchFromGitHub`
->
https://github.com/nixos/nixpkgs/blob/master/lib/customisation.nix#L134:C17